vmdk_plugin/drivers/vmdk/vmdk_driver.go

```diff
@@ -349,75 +425,43 @@ func (d *VolumeDriver) Path(r volume.Request) volume.Response {
 func (d *VolumeDriver) Mount(r volume.MountRequest) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Mounting volume ")
 
-	// If the volume is already mounted , just increase the refcount.
-	//
-	// Note: We are deliberately incrementing refcount first, before trying
-	// to do anything else. If Mount fails, Docker will send Unmount request,
-	// and we will happily decrement the refcount there, and will fail the unmount
-	// since the volume will have been never mounted.
-	// Note: for new keys, GO maps return zero value, so no need for if_exists.
-
-	refcnt := d.incrRefCount(r.Name) // save map traversal
-	log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
-	if refcnt > 1 {
-		log.WithFields(
-			log.Fields{"name": r.Name, "refcount": refcnt},
-		).Info("Already mounted, skipping mount. ")
-		return volume.Response{Mountpoint: getMountPoint(r.Name)}
-	}
-
-	// This is the first time we are asked to mount the volume, so comply
-	status, err := d.ops.Get(r.Name)
-
-	fstype := fs.FstypeDefault
-	isReadOnly := false
-	if err != nil {
-		d.decrRefCount(r.Name)
-		return volume.Response{Err: err.Error()}
-	}
-	// Check access type.
-	value, exists := status["access"].(string)
-	if !exists {
-		msg := fmt.Sprintf("Invalid access type for %s, assuming read-write access.", r.Name)
-		log.WithFields(log.Fields{"name": r.Name, "error": msg}).Error("")
-		isReadOnly = false
-	} else if value == "read-only" {
-		isReadOnly = true
-	}
-
-	// Check file system type.
-	value, exists = status["fstype"].(string)
-	if !exists {
-		msg := fmt.Sprintf("Invalid filesystem type for %s, assuming type as %s.",
-			r.Name, fstype)
-		log.WithFields(log.Fields{"name": r.Name, "error": msg}).Error("")
-		// Fail back to a default version that we can try with.
-		value = fs.FstypeDefault
+	// lock the state
+	d.refCounts.StateMtx.Lock()
+	defer d.refCounts.StateMtx.Unlock()
+	if d.refCounts.GetSuccess() != true {
+		// complete the mount process under the lock to avoid races
+		// dont do refcounting. Just mount
+		reponse := d.processMount(r, false)
+		d.refCounts.MarkDirty()
```

Comment:
Set MarkDirty before processMount so that it may save us some CPU cycles in discoverAndSync